Gh-79, 83 and 85 [Should be merged after #53]

rocky341 opened the pull request:
Pull request for issue #79, #83 and #85.

We are only going to include these 3 issues in this PR. Any further additions will be done in a separate PR so there isn't too much to review in one go. 

Is based off the code in #53 so that should be reviewed/merged before this one.

r32575 commented on gaffer-as-a-service/gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/model/StoreType.java:
```diff
@@ -13,18 +13,13 @@
-        return graph;
-    }
+public enum StoreType {
+    // TODO: rename to ACCUMULO_STORE & MAP_STORE
```
No longer needed. Should be a string instead.

rocky341 replied:
Deleted

r32575 commented:
I am on branch `gh-79-gaffer-maintainable` but this doesn't seem to have the changes made such as moving `common`. Any change you could either include them as separate pull requests or merge them as one branch in the future?

rocky341 commented:
> I am on branch `gh-79-gaffer-maintainable` but this doesn't seem to have the changes made such as moving `common`. Any change you could either include them as separate pull requests or merge them as one branch in the future?

Apologies for the confusion the correct branch is 'gh-79-gaffer-maintainability' which is the one corresponding to this PR. I have deleted `gh-79-gaffer-maintainable` as this was not worked on. You should find the gaas-common changes in this branch.

rocky341 commented on gaffer-as-a-service/gaas-common/findbugs-exclude.xml:
```diff
@@ -1,6 +1,6 @@
 <!--
-  ~ Copyright 2020 Crown Copyright
+  ~ Copyright 2016-2017 Crown Copyright
   ~
```
Updated

r32575 commented on gaffer-as-a-service/gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/factories/ProxyStoreType.java:
```diff
@@ -0,0 +1,68 @@
+        @Override
+        public GafferSpec build() {
+            final GafferSpec gafferSpec = super.build();
+            gafferSpec.putNestedObject(getDefaultProxyStoreProperties(graph.getProxyHost(), graph.getProxyContextRoot()), STORE_PROPERTIES_KEY);
```
Rather than having all the individual variables such as `proxyHost` and `proxyContextRoot` you could have a `storeProperties` `Map` containing all of these properties. You could add the method `setStoreProperties` to the `AbstractStoreTypeBuilder` and at `build` time you could search for the properties expected and if they don't exist throw an error such as `StoreTypeException`.

The `storeProperties` `Map` could be passed to the `rest` service `GaaSCreateRequestBody` taking the place of `proxyHost` and `proxyContextRoot`.  These values should be stored as key/values in the `storeProperties` `Map` instead. If you need to add values for new `StoreTypes` in the future then you could add them to the `storeProperties` rather than add new variables such as `proxyHost` and `proxyContextRoot`.

r32575 replied:
@rocky341, if you think this change is too much work then I can create a new ticket for it instead?

rocky341 replied:
@r32575 might be best to make this is a separate ticket if that's okay.

rocky341 replied:
@r32575 Actually will add to this PR

rocky341 replied:
This is now done

r32575 commented:
The `StoreTypes` could be moved to another package such as `stores` as they are not really `factories`.

r32575 commented:
>Apologies for the confusion the correct branch is gh-79-gaffer-maintainability which is the one corresponding to this PR and where you should find the gaas-common changes. I have deleted gh-79-gaffer-maintainable as this was not worked on.

Sorry, didn't realise there were multiple branches

r32575 commented on gaffer-as-a-service/gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/factories/GafferSpecBuilder.java:
```diff
@@ -0,0 +1,49 @@
+import static uk.gov.gchq.gaffer.gaas.util.Constants.INGRESS_UI_PATH_KEY;
+import static uk.gov.gchq.gaffer.gaas.util.Properties.INGRESS_SUFFIX;
+import static uk.gov.gchq.gaffer.gaas.util.Properties.NAMESPACE;
+public class GafferSpecBuilder {
```
Not sure if you need this `class`... Seems you could just build a `GafferSpec` object in the `AbstractStoreTypeBuilder.build()` instead and just store the `variables`.

r32575 replied:
Sorry, this goes against my sample code found here (https://github.com/gchq/gaffer-experimental/issues/79) but it would mean fewer classes.

rocky341 replied:
Done

rocky341 commented on gaffer-as-a-service/gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/factories/StoreTypeFactory.java:
```diff
@@ -0,0 +1,42 @@
+                return storeSpecManagers.getStoreSpecBuilder(graph);
+            }
+        }
+        throw new RuntimeException("StoreType is Invalid must be defined Valid Store Types supported are: federatedStore, accumuloStore, proxyStore and mapStore");
```
Done thank you

rocky341 commented on gaffer-as-a-service/gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/factories/StoreTypeFactory.java:
```diff
@@ -0,0 +1,42 @@
+    private final List<StoreType> storeTypeManager;
+
+    @Autowired
+    public StoreTypeFactory(final List<StoreType> storeTypeManager) {
```
Updated

rocky341 commented:
> The `StoreTypes` could be moved to another package such as `stores` as they are not really `factories`.

@r32575 Sorry that was meant to get moved after development. I have moved it now thanks. I have also moved `JsonObjectWrapper` into `util` from `factories`.

rocky341 commented on gaffer-as-a-service/gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/factories/GafferSpecBuilder.java:
```diff
@@ -0,0 +1,49 @@
+        gafferSpec.putNestedObject(this.graphId, GRAPH_ID_KEY);
+        gafferSpec.putNestedObject(description, DESCRIPTION_KEY);
+        gafferSpec.putNestedObject(graphId.toLowerCase() + "-" + NAMESPACE + "." + INGRESS_SUFFIX, INGRESS_HOST_KEY);
+        gafferSpec.putNestedObject("/rest", INGRESS_API_PATH_KEY);
```
`setStoreProperties` and `setSchema` have been added. 

For ingress such as `NAMESPACE` and `INGRESS_API_PATH_KEY` etc these are constants so don't think we need a setter for that.

rocky341 commented on gaffer-as-a-service/gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/factories/AccumuloStoreType.java:
```diff
@@ -0,0 +1,52 @@
+
+    private static final class AccumuloStoreSpecBuilder extends AbstractStoreTypeBuilder {
+
+        private AccumuloStoreSpecBuilder(final GaaSCreateRequestBody graph) {
```
Done

r32575 commented:
Getting a few merge conflicts that need to be resolved...

r32575 commented:
Just a minor thing: you haven't moved your test classes to separate folders in sync with the main classes

rocky341 commented:
@r32575 
> Just a minor thing: you haven't moved your test classes to separate folders in sync with the main classes
This has been updated thank you.


> Getting a few merge conflicts that need to be resolved...
Conflicts should be resolved now.
